src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -149,19 +149,84 @@ public static void DisposeFrames(RenderBatchBuilder batchBuilder, ArrayRange<Ren
             ref DiffContext diffContext,
             int oldStartIndex, int oldEndIndexExcl,
             int newStartIndex, int newEndIndexExcl)
+        {
+            // The overhead of the dictionary used by AppendAttributeDiffEntriesForRangeSlow is
+            // significant, so we want to try and do a merge-join if possible, but fall back to
+            // a hash-join if not. We'll do a merge join until we hit a case we can't handle and
+            // then fall back to the slow path.
+            //
+            // Also since duplicate attributes are not legal, we don't need to care about loops or
+            // the more complicated scenarios handled by AppendDiffEntriesForRange.
+            //
+            // We also assume that we won't see an attribute occur with different sequence numbers
+            // in the old and new sequences. It will be handled correct, but will generate a suboptimal
+            // diff.
+            var hasMoreOld = oldEndIndexExcl > oldStartIndex;
+            var hasMoreNew = newEndIndexExcl > newStartIndex;
+            var oldTree = diffContext.OldTree;
+            var newTree = diffContext.NewTree;
+
+            while (hasMoreOld || hasMoreNew)
+            {
+                var oldSeq = hasMoreOld ? oldTree[oldStartIndex].Sequence : int.MaxValue;
```

Comment:
I know it's only a tiny improvement, but to avoid the duplicated array lookups, we could have at the start of the loop:

```csharp
ref var oldItem = ref oldTree[oldStartIndex];
ref var newItem = ref newTree[newStartIndex];
```

It probably makes no difference for JIT-enabled runtimes, but may do for non-JIT ones that we're targeting.

Reply:
No, forget that, it makes it more complex given that we might have reached the end of one of the arrays already.